### PR TITLE
+htc #16834 apply uri-parsing-mode setting to header parsing as well

### LINKED ISSUE #16834: Clarify status of `akka.http.client.parsing.uri-parsing-mode`
since it is only used on the server-side.

Imported from spray/spray#555

/cc @sirthias

### PR DESCRIPTION
**Pull request by jrudolph**
Fixes #16834.

/cc @sirthias

### CODE REVIEW

**Comment by jrudolph** [comment]
Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3320/:

```
 11430 Test akka.stream.javadsl.FlowTest.mustBeAbleToRecover failed: java.lang.AssertionError: assertion failed: timeout (3 seconds) during expectMsg while waiting for 1, took 3.025 sec
 11431     at scala.Predef$.assert(Predef.scala:179)
 11432     at akka.testkit.TestKitBase$class.expectMsg_internal(TestKit.scala:338)
 11433     at akka.testkit.TestKitBase$class.expectMsg(TestKit.scala:315)
 11434     at akka.testkit.TestKit.expectMsg(TestKit.scala:718)
 11435     at akka.testkit.JavaTestKit.expectMsgEquals(JavaTestKit.java:378)
 11436     at akka.stream.javadsl.FlowTest.mustBeAbleToRecover(FlowTest.java:548)
 11437     ...
 12506 Failed: Total 867, Failed 1, Errors 0, Passed 866, Pending 22
 12507 Failed tests:
 12508 	akka.stream.javadsl.FlowTest
 15874 (akka-stream-tests-experimental/test:test) sbt.TestsFailedException: Tests unsuccessful
 15875 Total time: 847 s, completed Aug 4, 2015 12:28:38 PM
```

File: akka-http-core/src/main/scala/akka/http/impl/model/parser/HeaderParser.scala [hunk 1]
```diff
@@ -10,10 +10,12 @@ import akka.parboiled2._
 import akka.shapeless._
 import akka.http.scaladsl.model._
 
+private[http] case class HeaderParserSettings(uriParsingMode: Uri.ParsingMode = Uri.ParsingMode.Relaxed)
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/model/parser/HeaderParser.scala]: If we really need to wrap the `uriParsingMode`would it be better to cache the default instance in an `object`?

Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/model/parser/HeaderParser.scala]: The default code path is to construct this from `akka.http.server.parsing.uri-parsing-mode` in HttpHeaderParser. But you are right that I can probably get rid of the wrapping by changing the class into a trait.

Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/model/parser/HeaderParser.scala]: I changed the Settings into a trait similar to `HttpHeaderParser.Settings` which simplified things and also introduced a val with the default settings.

Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/model/parser/HeaderParser.scala]: Ah looks OK then. Just saw the default value in `HeaderParser` but overlooked that they are explicitly reused from the parser settings in `HttpHeaderParser`.

**Comment by jrudolph** [comment]
Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3321/:

```
 12746 Failed: Total 867, Failed 1, Errors 0, Passed 866, Pending 22
 12747 Failed tests:
 12748 	akka.stream.io.TcpSpec
 12755 Failed: Total 528, Failed 1, Errors 0, Passed 527, Ignored 467
 12756 Failed tests:
 12757 	akka.stream.tck.TransformProcessorTest
 16123 (akka-stream-tck-experimental/test:test) sbt.TestsFailedException: Tests unsuccessful
 16124 (akka-stream-tests-experimental/test:test) sbt.TestsFailedException: Tests unsuccessful
 16125 Total time: 904 s, completed Aug 4, 2015 2:09:28 PM
```

**Comment by drewhk** [comment]
More strange Jenkins errors but this PR is unrelated. LGTM

**Pull request merged**